lib/semian/sysv_shared_memory.rb

```diff
@@ -0,0 +1,50 @@
+module Semian
+  module SysVSharedMemory #:nodoc:
+    def self.included(base)
+      def base.do_with_sync(*names)
+        names.each do |name|
+          new_name = "#{name}_inner".freeze.to_sym
+          alias_method new_name, name
+          private new_name
+          define_method(name) do |*args, &block|
+            synchronize do
+              method(new_name).call(*args, &block)
+            end
+          end
+        end
+      end
+    end
```

Comment:
The things I still don't understand why you can't just do this for `#value` and reuse the implementation for the other ones, so you don't need this ugly meta programming that makes it hard to grep for and grok what's going on.